File: Domain/ValueObjects/Order/OrderStatus.cs
using OneOf;

namespace Domain.ValueObjects.Order;

public class OrderStatus : ValueObject
{
    public static OrderStatus Pending => new OrderStatus("Pending");
    public static OrderStatus Finished => new OrderStatus("Finished");
    public static readonly List<OrderStatus> ValidStatuses = [Pending, Finished];

    public string Name { get; }

    public OrderStatus(string name)
    {   
        Name = name;
    }

    public static OneOf<OrderStatus, string> CanCreate(string value)
    {
        var status = ValidStatuses.Find(status => status.Name == value);

        if (status is null)
        {
            return $"{value} is not a valid OrderStatus";
        }

        return status;
    }

    public static OrderStatus ExecuteCreate(string value)
    {
        var canCreateResult = CanCreate(value);

        if (canCreateResult.TryPickT1(out var error, out var status))
        {
            throw new Exception(error);
        }

        return status;
    }

    public static bool IsValid(string status)
    {
        return ValidStatuses.Exists(d => d.Name == status);
    }

    public override IEnumerable<object> GetEqualityComponents()
    {
        yield return Name;
    }
}

File: frontend/src/domain/valueObjects/Order/OrderStatus.ts
export default class OrderStatus {
    public static readonly PENDING = new OrderStatus('Pending');
    public static readonly FINISHED = new OrderStatus('Finished');

    private static readonly validStatuses = [
        OrderStatus.PENDING,
        OrderStatus.FINISHED
    ];

    private constructor(value: string) {
        this.value = value;
    }

    public static create(value: string) {
        if (value === OrderStatus.FINISHED.value) {
            return OrderStatus.FINISHED;
        } else if (value === OrderStatus.PENDING.value) {
            return OrderStatus.PENDING;
        } else {
            throw new Error(`${value} is not a valid OrderStatus`);
        }
    }

    public static isValid(status: string): boolean {
        return OrderStatus.validStatuses.some(validStatus => validStatus.value === status);
    }

    public value: string;
}